keps/sig-apps/20181220-statefulset-volume-expansion.md

```diff
@@ -0,0 +1,118 @@
+
+During the StatefulSet update process, the StatefulSet controller will detect an update to a `volumeClaimTemplate` by comparing the updated and current revision of the StatefulSet. This requires the `VolumeClaimTemplates` component of the StatefulSet to be recorded in the StatefulSet's `ControllerRevision` object. 
+
+While updating a pod, the StatefulSet controller will update a referenced persistent volume claim object if its storage request in the associated `volumeClaimTemplate` has been increased.
+
+Not all volumes support online control-plane expansion so this design aims to support both online and offline control-plane volume expansion.
+
+The functionality provided by this enhancement will be gated by the `StatefulSetVolumeExpansion` feature gate.
+
+For the initial version of this feature we will simply be expanding all volumes when the associated pod is offline. Below is the outline of how a resize will be propagated to each PVC:
+
+1. User updates a storage request in the `volumeClaimTemplates` component of a StatefulSet
+2. Apiserver validation validates that `StatefulSetVolumeExpansion` is enabled and that the storage request has not been decreased.
+3. The `PersistentVolumeClaimResize` admission controller verifies that the associated StorageClass for the `volumeClaimTemplate` being updated has `allowVolumeExpansion = true`
+4. Each PVC will be resized after the associated pod has been deleted by the StatefulSet controller. The controller will wait for expansion to complete before the pod is recreated. The controller will also wait for FileSystemExpansion to complete before concluding.
+
+In step 4 if volume expansion continues to fails we will document the following workaround for users. Note this is based off of the documented workaround to recover from a failing volume expansion.
+
+1. Delete the StatefulSet. Note PVCs and PVs are preserved.
+2. For the PV associated with the offending PVC, edit it to have the `Retain` reclaim policy.
+3. Delete the offending PVC.
+4. Recreate the PVC with the old spec and rebind to the same PV.
+5. Recreate the StatefulSet with the old spec.
+
+Note this issue can occur after some portion of the volumes have already been successfully resized. So after following the workaround the user can update the StatefulSet again to attempt to expand the remaining volumes. 
+
+#### Optimization for volumes that can be expanded online
+
+Note this is an optimization that may be added in later versions of this feature.
+
+Prerequisite: The StatefulSet controller must be able to determine if a volume supports online expansion.
+
+If a volume is deemed to support online expansion it will be expanded before the pod is terminated for the update. The StatefulSet controller will wait for the file system resize to complete on all such volumes before terminating the associated pod.
+
```

Comment:
Termination of the pod after a successful online expansion is unnecessary and incurs cost for the user.  e.g., for our Elasticsearch implementation, we migrate data off of a pod before terminating it in the pre-shutdown hook.   An additional sentence which describes the impediment to leaving the pod online would be helpful, so that consideration of a future enhancement will not need to retrace all the steps.

Reply:
Yes, the [`ExpandInUsePersistentVolumes` feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-gates-for-alpha-or-beta-features) allows expanding of a Volume in use by a Pod.

It works like a charm ;-)